### ISSUE TITLE
python3 and wx4

### ISSUE THREAD

**Issue by barcharcraz** [wrye-bash/wrye-bash#404]
Hello, I'm working on getting things running on python3 and wx4, hopefully it should make bash look a bit better and have less blocking UI operations and unicode bugs. It looks like a lot of code will be deleted in this port as I'm getting rid of all of bash's homegrown unicode handling.

I'm opening this issue so that anyone else who's also working on this can chime in.

**Comment by Utumno** [comment]
Well - python3 is not on the immediate plans - see #194 for my angry reaction back in 2015. Although the code is in much better shape now (and although @iaz3 hardly merited this angry reaction) python3 is still a way off. We have other priorities and having a merge changing basically every other line will mean end of business for some months and that's not an option.

Obviously changing to python 3 involves very careful scanning of the code (think only items vs iteritems, every single usage must be considered cause we may be modifying dictionaries for instance and blindly switching to p3 items will break code and that's just a random example). But there are other issues. What if we may want to use cython for the patcher ? That's p2 only.

Wx4 on the other hand is much more immediate goal. We are pretty close to wx3, but:

- I have remaining issues that nobody is looking at #15 - most importantly a wx.Locale traceback, see https://github.com/wrye-bash/wrye-bash/issues/15#issuecomment-322757698
- @nycz started a laudable but half baked attempt to _really_ solve this - as opposed to monkey patch. And really solving this is addressing #190 - that is really **wrapping** the wx library. Have a look at his branches here, namely nycz-wx-begone-DEV

So I am closing this as a duplicate. Let's continue the discussion in 15. I would rather have wx3 in after #190 is finally put to rest (_do have a look_ at linked commits there), fixed, tested, readme's updated, installer code patched, deployed and bugfixed. Then we can switch to wx4, easily and correctly.

**Comment by barcharcraz** [comment]
Thanks for pointing me there. I’ll clone nycz’s branch and work on that.

Also doesn’t cython support pay 3.3+?

**Comment by Utumno** [comment]
Be my guest :)
Keep in mind:

- there is one huge commit in there should be split so it's easier to review
- this is all experimental but to the right direction - the first few commits that are basically getting rid of the iewin dependency (an ancient issue) are more stable but still need review, packing and offering to the forum testers. Since webview is wx3 only we need also detailed instructions for people to upgrade in order to test
- nycz used `text` as an attribute later on, that's a nono so _if/when we get there_ we should change that. However this is later, no time for the sizers rewrite now let's focus to the binning of iewin/ wx3 conversion - there will be _surely_ issues cropping up. Let's continue in #15.

Re: cython: not sure had read at some point that's python2 only - anyway the point is that python3 needs thought (cython is just an idea in the back of my head I actually only spoken of it now). What I would consider an immediate goal for 308 (and a _huge_ one) would be **tests**. That would make all those library transitions much smoother. BAIN is far to complex to switch libraries without some tests - let alon the patcher...

**Comment by Metallicow** [comment]
@barcharcraz Any of the hacking on bash with wx4(project Phoenix) involved python2.7 as the base and should be a fairly easy finishing point to get there when the decision to go from "Classic" to "Phoenix" is officialy pushed through. As Utumno states Python3 support is another can of worms in itself as far as porting is concerned. ... So yea, if you want to be running/hacking with Phoenix(wx4) you should get that to a stable working point with Python2 before even ever considering any sort of major Python3 porting outside of getting code to work in a py2/py3 compatible fashion.

**Comment by Infernio** [comment]
Superseded by #460.
